Create rule S6775: All defaultProps should have non-required PropType (#3082)
SonarSource/rspec: rules/S6775/javascript/rule.adoc

== Why is this an issue?

> React Legacy APIs provides a way to define the default values and check the prop types at runtime. This rule verifies if a `defaultProp` definition does have a corresponding `propTypes` definition. If the one is missing it could be the result of errors in refactoring or a spelling mistake.

Reviewer: ```suggestion
React Legacy APIs provide a way to define the default values and check the prop types at runtime. This rule verifies if a `defaultProps` definition does have a corresponding `propTypes` definition. If it is missing, this could be the result of errors in refactoring or a spelling mistake.
```

Reviewer: Seems like something is missing here: "a way to define the default values". Default values for what?

Reviewer 2: Yes, you are right - replaced with `default values for props`

Reviewer: 👍

It is also an error if a `defaultProp` has `propType` that is marked as `isRequired`.

== How to fix it in PropTypes

[source,javascript,diff-id=1,diff-type=noncompliant]
----
function MyComponent({foo, bar}) {
  return <div>{foo}{bar}</div>;    
}

MyComponent.propTypes = {
  foo: React.PropTypes.string.isRequired,
};

MyComponent.defaultProps = {
  foo: "foo", // Noncompliant: foo is a required prop
  bar: "bar", // Noncompliant: bar propType is missing
};
----

> To fix the issue verify that each `defaultProp` has a corresponding `propType` definition and it is not marked as `isRequired`.

Reviewer: ```suggestion
To fix the issue, verify that each `defaultProp` has a corresponding `propType` definition and is not marked as `isRequired`.
```

[source,javascript,diff-id=1,diff-type=compliant]
----
function MyComponent({foo, bar}) {
  return <div>{foo}{bar}</div>;    
}

MyComponent.propTypes = {
  foo: React.PropTypes.string,
  bar: React.PropTypes.string,
};

MyComponent.defaultProps = {
  foo: "foo", 
  bar: "bar",
};
----

== How to fix it in TypeScript

[source,javascript,diff-id=2,diff-type=noncompliant]
----
type Props = {
  foo: string,
  bar?: string
}

function MyComponent({foo, bar}: Props) {
  return <div>{foo}{bar}</div>;    
}

MyComponent.defaultProps = {
  foo: "foo", // Noncompliant: foo is a required prop
  bar: "bar", 
};
----

> To fix the issue verify that each `defaultProp` has a corresponding type definition and it is marked as optional.

Reviewer: ```suggestion
To fix the issue, verify that each `defaultProp` has a corresponding type definition and is marked as optional.
```
